toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+	// The output buffer in memory will be used to re-create the initramfs.
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileInfo.Mode())
+
+	return
+}
+
+// Close flushes the archives and closes all initramfs resources
+func (i *InitramfsMount) Close() (err error) {
+	var bytesIO int
+
+	logger.Log.Debugf("Closing initramfs file '%s'", i.initramfsOutputFile.Name())
+
+	// Defer close calls to make sure we handle any errors, failing to
+	// close the file means we can't close the install root.
+	defer i.initramfsOutputFile.Close()
+	defer i.pgzWriter.Close()
+	defer i.cpioWriter.Close()
+
+	err = i.cpioWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+	err = i.pgzWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+
+	logger.Log.Debugf("Writing %d bytes to file", i.outputBuffer.Len())
+	bytesIO, err = i.initramfsOutputFile.Write(i.outputBuffer.Bytes())
+	if err != nil {
+		logger.Log.Errorf("Failed to write initramfs file: '%s'", err.Error())
+		return
+	}
+	logger.Log.Infof("Bytes writen to file: %d", bytesIO)
+
+	// Explicit call to fsync, archive corruption was occuring occasionally otherwise.
+	err = i.initramfsOutputFile.Sync()
+	if err != nil {
+		logger.Log.Errorf("Failed to sync initramfs file: '%s'", err.Error())
+		return
+	}
+
+	err = i.initramfsOutputFile.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+	return
+}
+
+// AddFileToInitramfs places a single file in the initramfs at the destination path.
+// - sourcePath: Path to file which is to be adde
+// - destPath: Final destination in the initramfs
+func (i *InitramfsMount) AddFileToInitramfs(sourcePath, destPath string) (err error) {
+	var bytesIO int64
+	fileInfo, err := os.Lstat(sourcePath)
+	file, err := os.Open(sourcePath)
+	if err != nil {
+		return
+	}
+	defer file.Close()
+
+	// Symlinks need to be resolved to their target file to be added to the cpio archive.
+	var link string
+	if fileInfo.Mode()&os.ModeSymlink != 0 {
+		link, err = os.Readlink(sourcePath)
+		if err != nil {
+			return
+		}
+
+		logger.Log.Debugf("--> Adding link: (%s) -> (%s)", sourcePath, link)
+	}
+
+	// Convert the OS header into a CPIO header
+	header, err := cpio.FileInfoHeader(fileInfo, link)
+	if err != nil {
+		return
+	}
+	header.Name = destPath
+
+	err = i.cpioWriter.WriteHeader(header)
+	if err != nil {
+		return
+	}
+
+	// Special files (unix sockets, directories, symlinks, ...) need to be handled differently
+	// since a simple byte transfer of the file's content into the CPIO archive can't be achieved.
+	if !fileInfo.Mode().IsRegular() {
+		// For a symlink the reported size will be the size (in bytes) of the link's target.
+		// Write this data into the archive.
+		if fileInfo.Mode()&os.ModeSymlink != 0 {
```

Comment:
>fileInfo.Mode()&os.ModeSymlink [](start = 5, length = 30)

Might be worth saving the result to a variable earlier on and not do this comparison twice.